Merge conflicts at the wrong moment could trigger unwanted API calls

e.g.

```
2016-09-26 17:16 dkt-github [ERROR] sync error: Merge conflict: /docker/*****/commit/e4460cb5bc78983018776f8f0c8e279a43e8720d/status/ci/datakit/osx-10.10/description: default
2016-09-26 17:16 dkt-github [INFO] Processing new entry
2016-09-26 17:16 dkt-github [INFO] API.set-pr {docker/****** 5322[e4460cb5bc78983018776f8f0c8e279a43e8720d] master open "******"}
```

In that case, the `API-set-pr` should not have happened, but as the branch is reverted the bridge become confused on what is a user request and what is just received from a webhook event.

--- a/bridge/github/src/datakit_github.ml
+++ b/bridge/github/src/datakit_github.ml
@@ -1485,6 +1485,13 @@ module Sync (API: API) (DK: Datakit_S.CLIENT) = struct
     close t.priv.tr >>= fun () ->
     close t.pub.tr
 
+  let safe_commit tr ~message =
+    DK.Transaction.commit tr ~message >>= function
+    | Ok ()   -> Lwt.return_unit
+    | Error e ->
+      Log.info (fun l -> l "Abort: %a" DK.pp_error e);
+      DK.Transaction.abort tr
+
   (* Merge the private branch back in the public branch. *)
   let merge t ~pub ~priv =
     assert (is_closed t);
@@ -1531,8 +1538,8 @@ module Sync (API: API) (DK: Datakit_S.CLIENT) = struct
             t.priv.name pp diff conflict_msg
         in
         Log.debug (fun l -> l "merge commit: %s" msg);
-        DK.Transaction.commit t.pub.tr ~message:msg >>*= fun () ->
-        abort t >>= fun () ->
+        safe_commit t.pub.tr ~message:msg >>= fun () ->
+        DK.Transaction.abort t.priv.tr  >>= fun () ->
         state "end-merge" ~old:(Some t) ~priv ~pub
 
   (** Sync *)
@@ -1554,7 +1561,7 @@ module Sync (API: API) (DK: Datakit_S.CLIENT) = struct
           | Ok ()   -> DK.Transaction.commit tr ~message:"Initial commit"
           | Error e ->
             DK.Transaction.abort tr >>= fun () ->
-            Lwt.fail_with @@ Fmt.strf "%a" DK.pp_error e
+            Lwt.fail_with @@ Fmt.strf "init_sync: %a" DK.pp_error e
         ) >>*= fun () ->
       with_head priv (DK.Branch.fast_forward pub)
     | Some pub_c, None  -> DK.Branch.fast_forward priv pub_c
@@ -1630,17 +1637,17 @@ module Sync (API: API) (DK: Datakit_S.CLIENT) = struct
     let priv_s, remove = Snapshot.prune priv_s in
     cleanup "sync" { remove; update = Some priv_s } t.priv.tr >>*= fun () ->
     DK.Transaction.diff t.priv.tr t.priv.head >>*= fun diff ->
-    (if diff = [] then DK.Transaction.abort t.priv.tr >>= ok else
+    (if diff = [] then DK.Transaction.abort t.priv.tr else
        let message = Fmt.strf "Sync with %a" Repo.Set.pp repos in
-       DK.Transaction.commit t.priv.tr ~message)
-    >>*= fun () ->
+       safe_commit t.priv.tr ~message)
+    >>= fun () ->
     DK.Transaction.abort t.pub.tr >>= fun () ->
     merge t ~pub ~priv >>*= fun t ->
     let pub_s, remove = Snapshot.prune t.pub.snapshot in
     cleanup "sync" { remove; update = Some pub_s } t.pub.tr >>*= fun () ->
     DK.Transaction.diff t.pub.tr t.pub.head >>*= fun diff ->
     (if diff = [] then ok t else
-       DK.Transaction.commit t.pub.tr ~message:"Prune" >>*= fun () ->
+       safe_commit t.pub.tr ~message:"Prune" >>= fun () ->
        DK.Transaction.abort t.priv.tr >>= fun () ->
        state "end-sync-repos" ~old:(Some t) ~pub ~priv)
 
@@ -1694,7 +1701,7 @@ module Sync (API: API) (DK: Datakit_S.CLIENT) = struct
           Fmt.strf "New webhook events\n\n%a"
             Fmt.(list ~sep:(unit "\n\n") Event.pp) events
         in
-        DK.Transaction.commit t.priv.tr ~message >>*= fun () ->
+        safe_commit t.priv.tr ~message >>= fun () ->
         DK.Transaction.abort t.pub.tr >>= fun () ->
         merge t ~pub ~priv
 

--- a/bridge/github/src/datakit_github_api.ml
+++ b/bridge/github/src/datakit_github_api.ml
@@ -103,10 +103,35 @@ module Ref = struct
   let to_list s = match String.cuts ~empty:false ~sep:"/" s with
     | "refs" :: l | l -> l
 
-  let of_gh repo r =
+  let commit_ref_of_tag ~token ~repo:{ Repo.user; repo } r =
+    let open Github.Monad in
+    assert (r.git_ref_obj.obj_ty = `Tag);
+    let sha = r.git_ref_obj.obj_sha in
+    let t =
+      Github.Repo.get_tag ~token ~user ~repo ~sha () >>~ fun t ->
+      (* FIXME: do we care about tags pointing to tags ?*)
+      Github.Monad.return { r with git_ref_obj = t.tag_obj }
+    in
+    Github.Monad.run t
+
+  open Lwt.Infix
+
+  let to_commit_ref ~token ~repo r =
+    match r.Github_t.git_ref_obj.obj_ty with
+    | `Blob
+    | `Tree   -> Lwt.return_none
+    | `Commit -> Lwt.return (Some r)
+    | `Tag    -> commit_ref_of_tag ~token ~repo r >|= fun r -> Some r
+
+  let of_gh_commit_ref ~repo r =
     assert (r.git_ref_obj.obj_ty = `Commit);
     let head = { Commit.repo; id = r.git_ref_obj.obj_sha } in
     { head; name = to_list r.git_ref_name }
+
+  let of_gh ~token ~repo r =
+    to_commit_ref ~token ~repo r >|= function
+    | Some r -> Some (of_gh_commit_ref ~repo r)
+    | None   -> None
 
   let of_event_hook repo r =
     let id = r.push_event_hook_after in
@@ -266,27 +291,15 @@ let prs token r =
 
 let refs token r =
   let { Repo.user; repo } = r in
-  let (>>~) = Github.Monad.(>>~) in
-  let ref_of_tag r =
-    assert (r.git_ref_obj.obj_ty = `Tag);
-    let sha = r.git_ref_obj.obj_sha in
-    Github.Repo.get_tag ~token ~user ~repo ~sha () >>~ fun t ->
-    Github.Monad.return { r with git_ref_obj = t.tag_obj }
-    (* FIXME: do we care about tags pointing to tags ?*)
-  in
   let refs ty =
-    let open Github.Monad in
     Github.Repo.refs ~ty ~token ~user ~repo ()
-    |> Github.Stream.to_list
-    >>= List.fold_left (fun acc r ->
-        match r.Github_t.git_ref_obj.obj_ty with
-        | `Blob
-        | `Tree   -> acc
-        | `Commit -> acc >|= fun acc -> r :: acc
-        | `Tag    -> ref_of_tag r >>= fun r -> acc >|= fun acc -> r :: acc
-      ) (Github.Monad.return [])
-    |> Github.Monad.map @@ List.map (Ref.of_gh r)
-    |> run
+    |>  Github.Stream.to_list
+    |>  Github.Monad.run
+    >>= Lwt_list.fold_left_s (fun acc ref ->
+        Ref.of_gh ~token ~repo:r ref >|= function
+        | None   -> acc
+        | Some r -> r :: acc
+      ) []
   in
   refs "heads" >>= fun heads ->
   refs "tags"  >|= fun tags  ->

--- a/src/client/datakit_S.mli
+++ b/src/client/datakit_S.mli
@@ -3,8 +3,8 @@
 open Result
 
 type stat = {
-  kind : [`File | `Dir | `Link | `Exec];
-  size : int64;
+  kind: [`File | `Dir | `Link | `Exec];
+  size: int64;
 }
 
 type status_state =
@@ -19,33 +19,33 @@ module type READABLE_TREE = sig
   type t
   type +'a or_error
 
-  val read : t -> Datakit_path.t ->
+  val read: t -> Datakit_path.t ->
     [`File of Cstruct.t | `Dir of string list | `Link of string] or_error Lwt.t
   (** [read t path] is the contents of the object at the [path]. *)
 
-  val stat : t -> Datakit_path.t -> stat option or_error Lwt.t
+  val stat: t -> Datakit_path.t -> stat option or_error Lwt.t
   (** [stat t path] is the metadata of the object at [path]. *)
 
-  val exists : t -> Datakit_path.t -> bool or_error Lwt.t
+  val exists: t -> Datakit_path.t -> bool or_error Lwt.t
   (** [exists t path] is [true] if [stat t path] isn't [None]. *)
 
-  val exists_file : t -> Datakit_path.t -> bool or_error Lwt.t
+  val exists_file: t -> Datakit_path.t -> bool or_error Lwt.t
   (** [exists_file t path] is similar to {!exists} but for files
       only. *)
 
-  val exists_dir : t -> Datakit_path.t -> bool or_error Lwt.t
+  val exists_dir: t -> Datakit_path.t -> bool or_error Lwt.t
   (** [exists_dir t path] is similar to {!exists} but for directories
       only. *)
 
-  val read_file : t -> Datakit_path.t -> Cstruct.t or_error Lwt.t
+  val read_file: t -> Datakit_path.t -> Cstruct.t or_error Lwt.t
   (** [read_file t path] resolves [path] to a file, or returns an
       error if it isn't a file. *)
 
-  val read_dir : t -> Datakit_path.t -> string list or_error Lwt.t
+  val read_dir: t -> Datakit_path.t -> string list or_error Lwt.t
   (** [read_dir t path] resolves [path] to a directory, or returns an
       error if it isn't one. *)
 
-  val read_link : t -> Datakit_path.t -> string or_error Lwt.t
+  val read_link: t -> Datakit_path.t -> string or_error Lwt.t
   (** [read_link t path] resolves [path] to a symlink, or returns an
       error if it isn't one. *)
 end
@@ -64,10 +64,10 @@ module type CLIENT = sig
   val error: ('a, unit, string, 'b or_error Lwt.t) format4 -> 'a
   (** [error] raises user-defined errors. *)
 
-  module Tree : READABLE_TREE with type 'a or_error := 'a or_error
+  module Tree: READABLE_TREE with type 'a or_error := 'a or_error
   (** A read-only tree of files, directories and symlinks. *)
 
-  module Commit : sig
+  module Commit: sig
     type t
     (** A [t] is an immutable commit in the database. *)
 
@@ -77,23 +77,23 @@ module type CLIENT = sig
     val compare: t -> t -> int
     (** [compare] compares commit IDs. *)
 
-    val tree : t -> Tree.t
+    val tree: t -> Tree.t
     (** [tree t] is the content of the commit. *)
 
-    val id : t -> string
+    val id: t -> string
     (** [id t] is the unique ID of this commit. *)
 
-    val message : t -> string or_error Lwt.t
+    val message: t -> string or_error Lwt.t
     (** [message t] is [t]'s log message. *)
 
-    val parents : t -> t list or_error Lwt.t
+    val parents: t -> t list or_error Lwt.t
     (** [parents t] is the list of [t]'s parent commits. *)
 
     val diff: t -> t -> Datakit_path.t diff list or_error Lwt.t
     (** [diff a b] returns the paths with differences between [a] and [b]. *)
   end
 
-  module Transaction : sig
+  module Transaction: sig
     (** All changes to a branch are made in transactions. When a
         transaction is committed, it is merged with the current
         contents of the branch. *)
@@ -104,71 +104,71 @@ module type CLIENT = sig
 
     (** {2 Writing} *)
 
-    val create_dir : t -> dir:Datakit_path.t -> string -> unit or_error Lwt.t
+    val create_dir: t -> dir:Datakit_path.t -> string -> unit or_error Lwt.t
     (** [create_dir t ~dir name] creates the directory [dir/name]. *)
 
-    val create_file : t -> dir:Datakit_path.t -> string -> ?executable:bool ->
+    val create_file: t -> dir:Datakit_path.t -> string -> ?executable:bool ->
       Cstruct.t -> unit or_error Lwt.t
     (** [create_file t ~dir name ?executable content] creates the file
         [dir/name]. *)
 
-    val create_symlink : t -> dir:Datakit_path.t -> string -> string ->
+    val create_symlink: t -> dir:Datakit_path.t -> string -> string ->
       unit or_error Lwt.t
     (** [create_symlink t ~dir name target] creates the symlink
         [dir/name]. *)
 
-    val replace_file : t -> dir:Datakit_path.t -> string -> Cstruct.t ->
+    val replace_file: t -> dir:Datakit_path.t -> string -> Cstruct.t ->
       unit or_error Lwt.t
     (** [replace_file t ~dir name new_content] changes the content of
         the existing file [dir/name]. *)
 
-    val create_or_replace_file : t -> dir:Datakit_path.t -> string -> Cstruct.t ->
+    val create_or_replace_file: t -> dir:Datakit_path.t -> string -> Cstruct.t ->
       unit or_error Lwt.t
     (** [create_or_replace_file t ~dir leaf content] uses either [create_file]
         or [replace_file] as appropriate to set the contents. *)
 
-    val set_executable : t -> Datakit_path.t -> bool -> unit or_error Lwt.t
+    val set_executable: t -> Datakit_path.t -> bool -> unit or_error Lwt.t
     (** [set_executable t path flag] marks the file at [path] as
         executable or not. *)
 
-    val remove : t -> Datakit_path.t -> unit or_error Lwt.t
+    val remove: t -> Datakit_path.t -> unit or_error Lwt.t
     (** [remove t path] removes [path]. If [path] is a directory then
         the entire subtree is removed. *)
 
-    val rename : t -> Datakit_path.t -> string -> unit or_error Lwt.t
+    val rename: t -> Datakit_path.t -> string -> unit or_error Lwt.t
     (** [rename t path new_name] changes the basename of [path] to
         [new_name].  Note: it is only possible to rename within a
         directory (this is a 9p limitation). *)
 
-    val truncate : t -> Datakit_path.t -> int64 -> unit or_error Lwt.t
+    val truncate: t -> Datakit_path.t -> int64 -> unit or_error Lwt.t
     (** [truncate t path length] sets the length of the file at [path]
         to [length].  If [length] is longer than the current length,
         the file is padded with zero bytes. *)
 
-    val make_dirs : t -> Datakit_path.t -> unit or_error Lwt.t
+    val make_dirs: t -> Datakit_path.t -> unit or_error Lwt.t
     (** [make_dirs t path] ensures that [path] exists and is a
         directory, creating it and any missing parents as
         necessary. *)
 
     (** {2 Finishing} *)
 
-    val commit : t -> message:string -> unit or_error Lwt.t
+    val commit: t -> message:string -> unit or_error Lwt.t
     (** [commit t ~message] creates a new commit with the given log
         message and the current contents of the transaction and then
         merges it into the branch from which the transaction was
         created.  The transaction cannot be used after calling
         this. *)
 
-    val abort : t -> unit Lwt.t
+    val abort: t -> unit Lwt.t
     (** [abort t] aborts the transaction without committing it.  The
         transaction cannot be used after calling this. *)
 
     (** {2 Merging and history} *)
 
     type merge_inputs = {
-      ours : Tree.t;
-      theirs : Tree.t;
-      base : Tree.t;
+      ours: Tree.t;
+      theirs: Tree.t;
+      base: Tree.t;
     }
     (** When performing a merge, these three directories can be used
         to calculate the final result.  [ours] is the previous
@@ -176,7 +176,7 @@ module type CLIENT = sig
         merged and [base] is a least common ancestor. If there is no
         common ancestor then [base] is an empty tree. *)
 
-    val merge : t -> Commit.t -> (merge_inputs * Datakit_path.t list) or_error Lwt.t
+    val merge: t -> Commit.t -> (merge_inputs * Datakit_path.t list) or_error Lwt.t
     (** [merge t commit] merges [commit] into the transaction. It
         performs any trivial merges it can and returns [(merge_inputs,
         conflicts)] to allow you to resolve the remaining ones. You
@@ -185,18 +185,18 @@ module type CLIENT = sig
         merges in one transaction, but the [merge_inputs] returned is
         only valid until the start of the next merge. *)
 
-    val parents : t -> Commit.t list or_error Lwt.t
+    val parents: t -> Commit.t list or_error Lwt.t
     (** [parents t] is the parents of the transaction (that is, the parents of
         the commit that would be generated if you committed now. *)
 
-    val set_parents : t -> Commit.t list -> unit or_error Lwt.t
+    val set_parents: t -> Commit.t list -> unit or_error Lwt.t
     (** [set_parents t new_parents] replaces the current list of
         parents.  Note that this does not perform a merge - it only
         affects the metadata.  Note also that [merge] automatically
         updates the parents, so it is not necessary to call it
         manually in that case. *)
 
-    val conflicts : t -> Datakit_path.t list or_error Lwt.t
+    val conflicts: t -> Datakit_path.t list or_error Lwt.t
     (** [conflicts t] returns the current list of paths that had merge
         conflicts and have not been written to since.  It is not
         possible to commit while this is non-empty. *)
@@ -211,26 +211,26 @@ module type CLIENT = sig
 
   end
 
-  module Branch : sig
+  module Branch: sig
     type t
     (** A [t] is a named pointer to a commit. *)
 
-    val name : t -> string
+    val name: t -> string
     (** [name t] is [t]'s name. *)
 
-    val remove : t -> unit or_error Lwt.t
+    val remove: t -> unit or_error Lwt.t
     (** [remove t] deletes branch [t]. If [t] does not exist, this
         does nothing. *)
 
-    val rename : t -> string -> unit or_error Lwt.t
+    val rename: t -> string -> unit or_error Lwt.t
     (** [rename t new_name] changes the name of this branch. It is an
         error if [new_name] already exists. *)
 
-    val head : t -> Commit.t option or_error Lwt.t
+    val head: t -> Commit.t option or_error Lwt.t
     (** [head t] is the current head of the branch, or [None] if it
         doesn't currently exist. *)
 
-    val wait_for_head : t -> ?switch:Lwt_switch.t ->
+    val wait_for_head: t -> ?switch:Lwt_switch.t ->
       (Commit.t option -> [`Finish of 'a | `Again | `Abort] or_error Lwt.t) ->
       [`Abort | `Finish of 'a] or_error Lwt.t
     (** [wait_for_head t fn] calls [fn commit] on the current
@@ -239,7 +239,7 @@ module type CLIENT = sig
         off the switch will make the wait return [`Abort] at the next
         opportunity. *)
 
-    val wait_for_path : t -> ?switch:Lwt_switch.t -> Datakit_path.t ->
+    val wait_for_path: t -> ?switch:Lwt_switch.t -> Datakit_path.t ->
       ([`File of Cstruct.t | `Dir of Tree.t
        | `Link of string | `Exec of Cstruct.t] option ->
        [`Finish of 'a | `Again | `Abort] or_error Lwt.t) ->
@@ -247,12 +247,12 @@ module type CLIENT = sig
     (** [wait_for_path] is similar to [wait_for_head], but waits for a
         particular sub-tree to change. *)
 
-    val fast_forward : t -> Commit.t -> unit or_error Lwt.t
+    val fast_forward: t -> Commit.t -> unit or_error Lwt.t
     (** [fast_forward t commit] changes [t] to point to [commit] if
         [t]'s head is an ancestor of [commit] (or returns an error if
         not). *)
 
-    val with_transaction : t -> (Transaction.t -> 'a or_error Lwt.t) ->
+    val with_transaction: t -> (Transaction.t -> 'a or_error Lwt.t) ->
       'a or_error Lwt.t
     (** [with_transaction t fn] is the result of applying [fn] to a
         new transaction on branch [t].  If the transaction has not
@@ -268,28 +268,28 @@ module type CLIENT = sig
 
   end
 
-  val branches : t -> string list or_error Lwt.t
+  val branches: t -> string list or_error Lwt.t
   (** [branches t] is the current set of branches. *)
 
-  val remove_branch : t -> string -> unit or_error Lwt.t
+  val remove_branch: t -> string -> unit or_error Lwt.t
   (** [remove_branch t name] removes the branch named [name] (unlike
       [Branch.remove], this method doesn't require creating the branch
       directory first). *)
 
-  val branch : t -> string -> Branch.t or_error Lwt.t
+  val branch: t -> string -> Branch.t or_error Lwt.t
   (** [branch t name] is the branch named [name] (which might not exist yet). *)
 
-  val commit : t -> string -> Commit.t
+  val commit: t -> string -> Commit.t
   (** [commit t hash] is the commit with ID [hash]. *)
 
-  val tree : t -> string -> Tree.t
+  val tree: t -> string -> Tree.t
   (** [tree t id] is the tree with ID [id]. *)
 
-  val fetch : t -> url:string -> branch:string -> Commit.t or_error Lwt.t
+  val fetch: t -> url:string -> branch:string -> Commit.t or_error Lwt.t
   (** [fetch t ~url ~branch] fetches the given remote branch and
       returns its head commit. *)
 
-  val disconnect : t -> unit Lwt.t
+  val disconnect: t -> unit Lwt.t
   (** [disconnect t] closes the connection. [t] cannot be used after
       this. *)
 end

